sql: run logic tests with fakedist-vec-disk

asubiotto opened the pull request:
This commit adds an explicit fakedist-vec-disk directive to some logic tests
and adds it to the default configurations to run if there are no directives.

Release note: None (testing change)

Most of the issues were related to `CatchUnsanitizedVectorizedRuntimeError`, we'll see if anything jumps out in CI as well.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45673)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
> Reviewed 24 of 24 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45673)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/execerror/error.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-M1Xfi8k0ZbRLlUVRJHx:-M1Xfi8k0ZbRLlUVRJHy:bno97u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1335bb49a0d0f5d1f149a52e8f15eb3564a152d2/pkg/sql/colexec/execerror/error.go#L64)):*
> ```Go
> }
> 
> // CatchUnsanitizedVectoriezdRuntimeError does the same as
> ```

nit: `s/riezd/rized/g`.
___
*[pkg/sql/colexec/execerror/error.go, line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-M1XgBII8NUe-SxlPT-t:-M1XgBII8NUe-SxlPT-u:b-a1tkca) ([raw file](https://github.com/cockroachdb/cockroach/blob/1335bb49a0d0f5d1f149a52e8f15eb3564a152d2/pkg/sql/colexec/execerror/error.go#L66)):*
> ```Go
> // CatchUnsanitizedVectoriezdRuntimeError does the same as
> // CatchSanitizedVectorizedRuntimeError but catches and returns panics as errors
> // from the vectorized engine without additional processing. Use this for
> ```

nit: I think it is now quite confusing which of the "catchers" to use in which scenarios. AFAIU, we want to use "sanitized" version only in the materializer that is the root of the vectorized flow, no? (And probably in `_test` files.)
___
*[pkg/sql/colflow/colrpc/outbox.go, line 227 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-M1Xgjzr-px5XXToitek:-M1Xgjzr-px5XXToitel:bzaqqsg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1335bb49a0d0f5d1f149a52e8f15eb3564a152d2/pkg/sql/colflow/colrpc/outbox.go#L227)):*
> ```Go
> 		}
> 
> 		if err := execerror.CatchSanitizedVectorizedRuntimeError(nextBatch); err != nil {
> ```

Should this be "unsanitized"?


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
It doesn't look like there's any reason why the `fakedist-vec-disk` should've caused this timeout. It seems it's `TestParallel` that is timing out and is tracked here: https://github.com/cockroachdb/cockroach/issues/38257. Trying out another CI run.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-:-M1bLRnu0RIid0ulmDXL:banxm27)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/execerror/error.go, line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-M1XgBII8NUe-SxlPT-t:-M1bIsxbAkMX1lbUvhAL:b232irq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1335bb49a0d0f5d1f149a52e8f15eb3564a152d2/pkg/sql/colexec/execerror/error.go#L66)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: I think it is now quite confusing which of the "catchers" to use in which scenarios. AFAIU, we want to use "sanitized" version only in the materializer that is the root of the vectorized flow, no? (And probably in `_test` files.)
</blockquote></details>

`CatchSanitizedRuntimeError` is when you are going to do something "special" with this error. `CatchUnsanitizedVectorizedRuntimeError` is for when you just want to pass this error along using `execerror.VectorizedInternalPanic`.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 227 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-M1Xgjzr-px5XXToitek:-M1bJ6J78GXx5RkSRg0U:bj4pncw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1335bb49a0d0f5d1f149a52e8f15eb3564a152d2/pkg/sql/colflow/colrpc/outbox.go#L227)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Should this be "unsanitized"?
</blockquote></details>

I thought so too, but a test was failing and it seems that the pgcode doesn't get properly propagated across the wire 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 6 of 6 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45673)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/sql/colexec/execerror/error.go, line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-M1XgBII8NUe-SxlPT-t:-M1bdZSQ7Lcs-_tG9plA:bl24mln) ([raw file](https://github.com/cockroachdb/cockroach/blob/1335bb49a0d0f5d1f149a52e8f15eb3564a152d2/pkg/sql/colexec/execerror/error.go#L66)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

`CatchSanitizedRuntimeError` is when you are going to do something "special" with this error. `CatchUnsanitizedVectorizedRuntimeError` is for when you just want to pass this error along using `execerror.VectorizedInternalPanic`.
</blockquote></details>

Ok, the clarifications are helpful.
___
*[pkg/sql/colflow/colrpc/outbox.go, line 227 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-M1Xgjzr-px5XXToitek:-M1bdzVcBBeddPv4nTFX:b-vbsm1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1335bb49a0d0f5d1f149a52e8f15eb3564a152d2/pkg/sql/colflow/colrpc/outbox.go#L227)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I thought so too, but a test was failing and it seems that the pgcode doesn't get properly propagated across the wire 
</blockquote></details>

Hm, maybe the fact the we're wrapping the errors with `nonVectorizedInternalError` causing the not propagation across the wire. I thought that @knz's work on `errors` library made the errors "round-trippable".


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45673)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colflow/colrpc/outbox.go, line 227 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-M1Xgjzr-px5XXToitek:-M1eDfn348i35hgLe-Kb:b-n0dhly) ([raw file](https://github.com/cockroachdb/cockroach/blob/1335bb49a0d0f5d1f149a52e8f15eb3564a152d2/pkg/sql/colflow/colrpc/outbox.go#L227)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Hm, maybe the fact the we're wrapping the errors with `nonVectorizedInternalError` causing the not propagation across the wire. I thought that @knz's work on `errors` library made the errors "round-trippable".
</blockquote></details>

I just checked the code - of course this error is broken

1) (I think  but am not sure) `notVectorizedInternalError` does not properly implement the `error` interface - it should have its own `Error` method instead of naively embedding an `error` object.

2) `notVectorizedInternalError` is not implementing `Cause()` / `Unwrap()` so the errors library doesn't know what to do with it

3) if you want to have it preserved properly across the wire, you need to have a deserialization function and use `errors.RegisterWrapperDecoder(...)`

4) there's something fishy going in `SanitizeVectorizedRuntimeError`, the following code is quite incorrect: (and quite unscrutable)

```go
    doNotAnnotate ≔ false
    if nvie, ok ≔ err.(*notVectorizedInternalError); ok {
      // A notVectorizedInternalError was not caused by the
      // vectorized engine and represents an error that we don't
      // want to annotate in case it doesn't have a valid PG code.
      doNotAnnotate = true
      // We want to unwrap notVectorizedInternalError so that in case
      // the original error does have a valid PG code, the code is
      // correctly propagated.
      retErr = nvie.err
    }
    if code ≔ pgerror.GetPGCode(retErr); !doNotAnnotate && code ≡ pgcode.Uncategorized {
      // Any error without a code already is "surprising" and
      // needs to be annotated to indicate that it was
      // unexpected.
      retErr = errors.AssertionFailedf("unexpected error from the vectorized runtime: %+v", retErr)
    }

```

Try this instead:

```go
rejectErrorWithoutCode := true
var nvie *notVectorizedInternalError
if errors.As(err, &nvie) {
      // A notVectorizedInternalError was not caused by the
      // vectorized engine and represents an error that we don't
      // want to annotate in case it doesn't have a valid PG code.
      rejectErrorWithoutCode = false
}
 if code ≔ pgerror.GetPGCode(err);rejectErrorWithoutCode && code ≡ pgcode.Uncategorized {
      // Any error without a code already is "surprising" and
      // needs to be annotated to indicate that it was
      // unexpected.
      err = errors.NewAssertionFailureWithWrappedErrf(err, "unexpected error from the vectorized runtime")
    }
return err
```

please note:
- no need for retErr
- no need to unwrap the notVectorizedInternalError (this assume it implements Cause/Unwrap properly!! you need to do this in any case)
- errors.As
- don't lose the original error object when creating the assertion failure


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
Reworked with @knz's suggestion which I think is much nicer (thanks!).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-:-M1kL1su3KnBC-Xzb_kO:bx830di)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colflow/colrpc/outbox.go, line 227 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-M1Xgjzr-px5XXToitek:-M1bfOj59HDc_53bCpsI:b-jc75hg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1335bb49a0d0f5d1f149a52e8f15eb3564a152d2/pkg/sql/colflow/colrpc/outbox.go#L227)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I just checked the code - of course this error is broken

1) (I think  but am not sure) `notVectorizedInternalError` does not properly implement the `error` interface - it should have its own `Error` method instead of naively embedding an `error` object.

2) `notVectorizedInternalError` is not implementing `Cause()` / `Unwrap()` so the errors library doesn't know what to do with it

3) if you want to have it preserved properly across the wire, you need to have a deserialization function and use `errors.RegisterWrapperDecoder(...)`

4) there's something fishy going in `SanitizeVectorizedRuntimeError`, the following code is quite incorrect: (and quite unscrutable)

```go
    doNotAnnotate ≔ false
    if nvie, ok ≔ err.(*notVectorizedInternalError); ok {
      // A notVectorizedInternalError was not caused by the
      // vectorized engine and represents an error that we don't
      // want to annotate in case it doesn't have a valid PG code.
      doNotAnnotate = true
      // We want to unwrap notVectorizedInternalError so that in case
      // the original error does have a valid PG code, the code is
      // correctly propagated.
      retErr = nvie.err
    }
    if code ≔ pgerror.GetPGCode(retErr); !doNotAnnotate && code ≡ pgcode.Uncategorized {
      // Any error without a code already is "surprising" and
      // needs to be annotated to indicate that it was
      // unexpected.
      retErr = errors.AssertionFailedf("unexpected error from the vectorized runtime: %+v", retErr)
    }

```

Try this instead:

```go
rejectErrorWithoutCode := true
var nvie *notVectorizedInternalError
if errors.As(err, &nvie) {
      // A notVectorizedInternalError was not caused by the
      // vectorized engine and represents an error that we don't
      // want to annotate in case it doesn't have a valid PG code.
      rejectErrorWithoutCode = false
}
 if code ≔ pgerror.GetPGCode(err);rejectErrorWithoutCode && code ≡ pgcode.Uncategorized {
      // Any error without a code already is "surprising" and
      // needs to be annotated to indicate that it was
      // unexpected.
      err = errors.NewAssertionFailureWithWrappedErrf(err, "unexpected error from the vectorized runtime")
    }
return err
```

please note:
- no need for retErr
- no need to unwrap the notVectorizedInternalError (this assume it implements Cause/Unwrap properly!! you need to do this in any case)
- errors.As
- don't lose the original error object when creating the assertion failure

</blockquote></details>

Reworked properly and also changed the code to avoid the indentation that was present. @knz I'm curious about a couple of things:
1) Why is it not enough to embed an error to implement the `Error` interface? My impression is that it is.
2) What is the difference between `errors.As` and `errors.Is`?
I think we do need `retErr` though because this is a recovery running in a `defer`


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
Also note that I've removed hard limits from buffering operators that do not have a spilling strategy. They are still hooked up to the monitor but won't error out if we set the work memory limit to 1 in logic tests. This is equivalent to what the row execution engine does.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-:-M1kL9Uo3F2idwt9quBt:bme1ay8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45673)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colflow/colrpc/outbox.go, line 227 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-M1Xgjzr-px5XXToitek:-M1kM2aBAzX9eWULg4_j:byr4298) ([raw file](https://github.com/cockroachdb/cockroach/blob/1335bb49a0d0f5d1f149a52e8f15eb3564a152d2/pkg/sql/colflow/colrpc/outbox.go#L227)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Reworked properly and also changed the code to avoid the indentation that was present. @knz I'm curious about a couple of things:
1) Why is it not enough to embed an error to implement the `Error` interface? My impression is that it is.
2) What is the difference between `errors.As` and `errors.Is`?
I think we do need `retErr` though because this is a recovery running in a `defer`
</blockquote></details>

1. because by implementing the way you did your wrapper type is not really a wrapper. The error library will not detect it's a wrapper because it does not implement Cause/Unwrap. It won't be able to see its type.
2. errors.As is a type comparison (and retrieves the object that matches the requested type). errors.Is is a value comparison.

I'll defer (hah!) to you regarding retErr.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45673)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colflow/colrpc/outbox.go, line 227 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-M1Xgjzr-px5XXToitek:-M1kO0F3E-wFxN7LD5hh:bnkrvwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1335bb49a0d0f5d1f149a52e8f15eb3564a152d2/pkg/sql/colflow/colrpc/outbox.go#L227)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1. because by implementing the way you did your wrapper type is not really a wrapper. The error library will not detect it's a wrapper because it does not implement Cause/Unwrap. It won't be able to see its type.
2. errors.As is a type comparison (and retrieves the object that matches the requested type). errors.Is is a value comparison.

I'll defer (hah!) to you regarding retErr.
</blockquote></details>

I _think_ it's fine to embed the `error` object _and_ implement Cause/Unwrap though. 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 3 of 27 files at r5, 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45673)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colexec/execerror/error.go, line 150 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-M1kiFLJ4C726dmtagdA:-M1kiFLK5XAvRlM6k7i-:b-b9608) ([raw file](https://github.com/cockroachdb/cockroach/blob/50edb4605fbf144ba1342d3f37d7afc5c4b760e9/pkg/sql/colexec/execerror/error.go#L150)):*
> ```Go
> // notVectorizedInternalError will be returned to the client not as an
> // "internal error" and without the stack trace.
> type notVectorizedInternalError struct {
> ```

nit: seems like `notVectorizedInternalError` now implements some interface, we should add a thing to check that (i.e. `var _ <interface> = ...`).


<!-- Sent from Reviewable.io -->

Commit:
sql: run logic tests with fakedist-vec-disk

This commit adds an explicit fakedist-vec-disk directive to some logic tests
and adds it to the default configurations to run if there are no directives.

Release note: None (testing change)

Commit:
execerror: use errors package

We previously would unwrap a notVectorizedInternalError when first catching it,
which mean that we would lose that information and be unable to propagate its
underlying cause without annotating it if caught again (since we only check
for a *notVectorizedInternal wrapper). Instead, this patch changes the
CatchVectorizedRuntimeError function to leave the error object unwrapped when
doing its annotation. *notVectorizedInternalError is now enhanced so that it
can be sent over the wire unwrapped and propagated as many times as necessary
through CatchVectorizedRuntimeError without losing information.

Release note (bug fix): expected errors from the vectorized execution engine
are no longer mistakenly annotated as unexpected errors.

asubiotto commented:
I added `execerror/error.go` to the whitelist for `proto.Message` since it needs to import that package to provide decoding.

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 6 of 6 files at r7, 4 of 4 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45673#-:-M1lBQRe7cwJKyeDpUW8:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
I think we should backport the fix of `colexec/error.go` to 19.2.
